### PR TITLE
Further rule tuning and optic-ci command fixes

### PR DESCRIPTION
**Pull request by acunniffe**
While debugging https://github.com/snyk/sweater-comb/issues/122 we realized there's further tuning required. Specifically for the breaking change-type rules. Sometimes `added` should run if the parent was added, sometimes it shouldn't. This has more to do with the structure of the OpenAPI document than the API domain. For each kind of rule, it should behave differently. 

Also fixes 
https://github.com/snyk/sweater-comb/issues/124
https://github.com/snyk/sweater-comb/issues/121

### CODE REVIEW

File: src/dsl.ts [hunk 1]
```diff
@@ -397,6 +399,7 @@ export class SnykApiCheckDsl implements ApiCheckDsl {
         (location) => dsl.getContext(location),
         (...items) => dsl.checks.push(...items),
         (pointer: string) => jsonPointerHelpers.get(dsl.nextJsonLike, pointer),
+        (parent) => parent.location.kind === OpenApiKind.Response,
```
Comment (acunniffe) [inline src/dsl.ts:402]: ignore added if response was added at the same time

File: src/dsl.ts [hunk 2]
```diff
@@ -362,6 +363,7 @@ export class SnykApiCheckDsl implements ApiCheckDsl {
         (location) => dsl.getContext(location),
         (...items) => dsl.checks.push(...items),
         (pointer: string) => jsonPointerHelpers.get(dsl.nextJsonLike, pointer),
+        (parent) => parent.location.kind === OpenApiKind.Operation,
```
Comment (acunniffe) [inline src/dsl.ts:366]: skip added if operation was added at the same time

File: src/dsl.ts [hunk 3]
```diff
@@ -334,6 +334,7 @@ export class SnykApiCheckDsl implements ApiCheckDsl {
...
```
Comment (acunniffe) [inline src/dsl.ts:337]: skip added if operation was added at the same time

File: src/dsl.ts [hunk 4]
```diff
@@ -470,6 +473,11 @@ export class SnykApiCheckDsl implements ApiCheckDsl {
       (location) => dsl.getContext(location),
       (...items) => dsl.checks.push(...items),
       (pointer: string) => jsonPointerHelpers.get(dsl.nextJsonLike, pointer),
+      (parent) =>
+        !(
+          parent.location.kind === OpenApiKind.Body ||
+          parent.location.kind === OpenApiKind.Field
+        ),
     );
```
Comment (acunniffe) [inline src/dsl.ts:481]: skip all nested added for properties so breaking change rules don't apply when request body or response body is added

**Comment by acunniffe** [comment]
This needs a little bit more work re: failing tests. Working on a robust fix now.

**Comment by acunniffe** [comment]
Ok @cmars  -- I finally feel good about this. I realized, it's important for the rules themselves to know about their parents to get the right level of control we'll need for all the current / future cases here. See inline comments

File: src/rulesets/operations.ts [hunk 5]
```diff
@@ -169,6 +169,7 @@ export const rules = {
     request.queryParameter.added.must(
       "not be required",
       (queryParameter, context, docs) => {
+        if (context.operationAdded) return;
```
Comment (acunniffe) [inline src/rulesets/operations.ts:172]: for rules about required, they can be required only if the operation was also added at the same time (there to start).

File: src/rulesets/properties.ts [hunk 6]
```diff
@@ -96,6 +96,7 @@ export const rules = {
     bodyProperties,
   }: SnykApiCheckDsl) => {
     bodyProperties.added.must("not be required", (property, context, docs) => {
+      if (context.bodyAdded) return;
```
Comment (acunniffe) [inline src/rulesets/properties.ts:99]: for rules about required, they can be required only if the body was also added at the same time (there to start).

**Comment by cmars** [comment]
Seeing an issue with the e2e test fixtures in this branch at 3a3f04add6c8c6b22cde48dd6276d7978c1b2759. For example:
```
ksnyk:sweater-comb cmars$ yarn run compare --from ./end-end-tests/api-standards/resources/thing/2021-11-10/000-baseline.yaml ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml 
yarn run v1.22.11
$ yarn ts-node src/index.ts compare --from ./end-end-tests/api-standards/resources/thing/2021-11-10/000-baseline.yaml ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml
$ /Users/cmars/Projects/sweater-comb/node_modules/.bin/ts-node src/index.ts compare --from ./end-end-tests/api-standards/resources/thing/2021-11-10/000-baseline.yaml ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml
Loading specifications for comparison:
Current specification: done
Next specification: done
running rules...
Error running rules: "Cannot read property 'index' of undefined"
error Command failed with exit code 1.
```

**Comment by acunniffe** [comment]
Looks like we're back to compare issues, I'll sync with the team this AM and try to get a better understanding of what broke and why we've had such trouble making this reliable the last week. I imagine it's because the rules throw, and managing the lifecycle of the process, esp with multiple compares running adds a lot of logic and try/catch branches. 

On the positive, that same test case is in the code tests and passes: https://github.com/acunniffe/sweater-comb/blob/3a3f04add6c8c6b22cde48dd6276d7978c1b2759/src/rulesets/tests/end-end.test.ts#L39

**Comment by acunniffe** [comment]
Hey @cmars  -- so we found a small issue w/ the compare function, but it also looks like you ran this without `--to`. Can you try this? 
```
yarn ts-node src/index.ts compare --from ./end-end-tests/api-standards/resources/thing/2021-11-10/000-baseline.yaml --to ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml
```

**Comment by cmars** [comment]
> Hey @cmars -- so we found a small issue w/ the compare function, but the team realized it looks like you ran this without --to.

Good catch, it was probably too late for me...

Ok, with a night's rest and caffeine, here's where we're at:

```
yarn run v1.22.11
$ /Users/cmars/Projects/sweater-comb/node_modules/.bin/ts-node src/index.ts compare --from ./end-end-tests/api-standards/resources/thing/2021-11-10/000-baseline.yaml --to ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml --context { '"changeDate":' '"2021-11-11",' '"changeResource":' '"thing",' '"changeVersion":' { '"date":' '"2021-11-10",' '"stability":' '"experimental"' }, '"resourceVersionReleases":' {} }
Loading specifications for comparison:
Current specification: loading...
Next specification: loading...
[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: loading...
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...


  FAIL   Entire Resource
  x api lifeycle:  must lifeycle rules have to be followed
    Cannot read property 'date' of undefined
    Read more in our API guide (​https://github.com/snyk/sweater-comb/blob/main/docs/version.md​)
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)
  x api lifeycle:  must follow sunset rules
    Cannot read property 'stability' of undefined
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)

```

The above seems to be a bug in the lifecycle rules. Or is it a problem with my context object. I seem to get these errors with and without a context.

```

  PASS   Resource Document


  FAIL   This Specification
  x requirement  must oas3-parameter-description
    Parameter objects must have "description".
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:83:4085)

```

A legit rule failure, which wasn't a failure intended by the fixture. Good catch `optic-ci`, I'll update my test!

```

  PASS  DELETE /orgs/{org_id}/thing/{thing_id}


  PASS  GET /orgs/{org_id}/thing


  FAIL  GET /orgs/{org_id}/thing/{thing_id}
  x requirement for operation: GET /orgs/{org_id}/thing/{thing_id} must use UUID for org_id or group_id
    expected operation /orgs/{org_id}/thing/{thing_id} get parameter org_id to use format UUID: expected undefined to equal 'uuid'
    Read more in our API guide (​https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources​)
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:73:3725)
  x added query-parameter: org_id must not be required
    expected true to be false
    Read more in our API guide (​https://github.com/snyk/sweater-comb/blob/main/docs/version.md#breaking-changes​)
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:83:4085)

```

More legit failures. The only intended failure here was the "breaking change", I need to add the format to the fixture.

```
  PASS  PATCH /orgs/{org_id}/thing/{thing_id}


  PASS  POST /orgs/{org_id}/thing





429 checks passed
5 checks failed
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by cmars** [comment]
Also note the lifecycle rule error when introducing a "new" file (other rule failures in this spec file are legit as noted above):

```
yarn run v1.22.11
$ /Users/cmars/Projects/sweater-comb/node_modules/.bin/ts-node src/index.ts compare --to ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml --context { '"changeDate":' '"2021-11-11",' '"changeResource":' '"thing",' '"changeVersion":' { '"date":' '"2021-11-10",' '"stability":' '"experimental"' }, '"resourceVersionReleases":' {} }
Loading specifications for comparison:
Current specification: loading...
Next specification: loading...
[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: loading...
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...


  FAIL   Entire Resource
  x api lifeycle:  must lifeycle rules have to be followed
    Cannot read property 'date' of undefined
    Read more in our API guide (​https://github.com/snyk/sweater-comb/blob/main/docs/version.md​)
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)
  x api lifeycle:  must follow sunset rules
    Cannot read property 'stability' of undefined
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)


  PASS   Resource Document


  FAIL   This Specification
  x requirement  must oas3-parameter-description
    Parameter objects must have "description".
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:83:4085)


  PASS  DELETE /orgs/{org_id}/thing/{thing_id}


  PASS  GET /orgs/{org_id}/thing


  FAIL  GET /orgs/{org_id}/thing/{thing_id}
  x requirement for operation: GET /orgs/{org_id}/thing/{thing_id} must use UUID for org_id or group_id
    expected operation /orgs/{org_id}/thing/{thing_id} get parameter org_id to use format UUID: expected undefined to equal 'uuid'
    Read more in our API guide (​https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources​)
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:73:3725)


  PASS  PATCH /orgs/{org_id}/thing/{thing_id}


  PASS  POST /orgs/{org_id}/thing





2101 checks passed
4 checks failed
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by acunniffe** [comment]
> The above seems to be a bug in the lifecycle rules. Or is it a problem with my context object? I seem to get these errors with and without a context.

That is not a valid context object, it isn't escaped properly. We really need better feedback on the parsing side to help
```
--context { '"changeDate"...
```
^ not escaped

Again for the failed lifecycle rule in most recent comment
```
--context { '"changeDate":
```
^ not escaped

**Comment by cmars** [comment]
Good catch on the shell escaping! I think this run should be escaped properly (needed to double-quote the `"$(cat /path/to/context.json)"`). Same result though:

```
yarn run v1.22.11
$ /Users/cmars/Projects/sweater-comb/node_modules/.bin/ts-node src/index.ts compare --to ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml --context '{
  "changeDate": "2021-11-11",
  "changeResource": "thing",
  "changeVersion": {
    "date": "2021-11-10",
    "stability": "experimental"
  },
  "resourceVersionReleases": {}
}'
Loading specifications for comparison:
Current specification: loading...
Next specification: loading...
[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: loading...
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...


  FAIL   Entire Resource
  x api lifeycle:  must lifeycle rules have to be followed
    Cannot read property 'date' of undefined
    Read more in our API guide (​https://github.com/snyk/sweater-comb/blob/main/docs/version.md​)
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)
  x api lifeycle:  must follow sunset rules
    Cannot read property 'stability' of undefined
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)


  PASS   Resource Document


  PASS   This Specification


  PASS  DELETE /orgs/{org_id}/thing/{thing_id}


  PASS  GET /orgs/{org_id}/thing


  PASS  GET /orgs/{org_id}/thing/{thing_id}


  PASS  PATCH /orgs/{org_id}/thing/{thing_id}


  PASS  POST /orgs/{org_id}/thing





2102 checks passed
2 checks failed
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: src/rulesets/api-lifeycle.ts [hunk 7]
```diff
@@ -3,16 +3,6 @@ import { expect } from "chai";
...
-  example: ({ checkApiContext, responses }: SnykApiCheckDsl) => {
-    checkApiContext.must(
-      "lifeycle rules have to be followed",
-      (context, docs) => {
-        docs.includeDocsLink(links.versioning.main);
-        context.changeVersion.date;
-      },
-    );
-  },
-
```
Comment (acunniffe) [inline src/rulesets/api-lifeycle.ts:15]: it was failing here, and this did nothing

**Comment by acunniffe** [comment]
Ok removed that failing check (it wasn't expecting anything). 

I also debugged this with the team and realized yes, you're right! And I'm sorry to have told you otherwise. The `--context` for compare is broken and is passing around a string, the bulk-compare one works. 

I'm sorry for the confusion here. It's being cleaned up now.

**Comment by cmars** [comment]
That gets us a little closer. Just one erroring rule now failing a "completely new spec" scenario:

```
yarn run v1.22.11
$ /Users/cmars/Projects/sweater-comb/node_modules/.bin/ts-node src/index.ts compare --to ./end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml --context '{
  "changeDate": "2021-11-11",
  "changeResource": "thing",
  "changeVersion": {
    "date": "2021-11-10",
    "stability": "experimental"
  },
  "resourceVersionReleases": {}
}'
Loading specifications for comparison:
Current specification: loading...
Next specification: loading...
[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: loading...
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...
[2K[1A[2K[1A[2K[1A[2K[1A[2K[GLoading specifications for comparison:
Current specification: done
Next specification: done
running rules...


  FAIL   Entire Resource
  x api lifeycle:  must follow sunset rules
    Cannot read property 'stability' of undefined
    at (/Users/cmars/Projects/sweater-comb/end-end-tests/api-standards/resources/thing/2021-11-10/001-fail-breaking-param-change.yaml:4:125)


  PASS   Resource Document


  PASS   This Specification


  PASS  DELETE /orgs/{org_id}/thing/{thing_id}


  PASS  GET /orgs/{org_id}/thing


  PASS  GET /orgs/{org_id}/thing/{thing_id}


  PASS  PATCH /orgs/{org_id}/thing/{thing_id}


  PASS  POST /orgs/{org_id}/thing





2102 checks passed
1 checks failed
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by cmars** [comment]
Nice to have: `--context` accepts a path to a context.json file, rather than a literal JSON string. That'd help avoid escaping issues w/shells.

**Pull request merged**